File: algorithms/308.Range_Sum_Query_2D_-_Mutable/Range_Sum_Query_2D_-_Mutable.cc
<<<<<<< HEAD
class NumMatrix {
public:
    NumMatrix(vector<vector<int>> &matrix) {
        if(matrix.size() == 0)
            return ;
        mtx.resize(matrix.size() + 1);
        sumMtx.resize(matrix.size() + 1);
        for(int i = 0; i <= matrix.size(); ++ i){
            vector<int> tmp(matrix[0].size() + 1), sumTmp(matrix[0].size() + 1);
            mtx[i] = tmp;
            sumMtx[i] = sumTmp;
        }
        for(int i = 0; i < matrix.size(); ++ i){
            for(int j = 0; j < matrix[0].size(); ++ j){
                mtx[i][j] = matrix[i][j];
                add(i + 1, j + 1, matrix.size(), matrix[0].size(), mtx[i][j]);
            }
        }
    }

    void update(int row, int col, int val) {
        if(mtx.size() == 0)
            return ;
        if(mtx[row][col] == val)
            return ;
        int delta = val - mtx[row][col];
        add(row + 1, col + 1, mtx.size() - 1, mtx[0].size(), delta);
        mtx[row][col] = val;
    }

    int sumRegion(int row1, int col1, int row2, int col2) {
        if(mtx.size() == 0)
            return 0;
        if(!row1 && !col1)
            return sum(row2, col2, mtx[0].size());
        else if(!row1)
            return sum(row2, col2, mtx[0].size()) - sum(row2, col1 - 1, mtx[0].size());
        else if(!col1)
            return sum(row2, col2, mtx[0].size()) - sum(row1 - 1, col2, mtx[0].size());
        else
            return sum(row2, col2, mtx[0].size()) - sum(row2, col1 - 1, mtx[0].size()) - sum(row1 - 1, col2, mtx[0].size()) + sum(row1 - 1, col1 - 1, mtx[0].size());
    }
private:
    void add(int startRow, int startCol, int endRow, int endCol, int val){
        for(int i = startRow; i <= endRow; i += lowbit(i)){
            sumMtx[i][startCol] += val;
        }
    /*    for(int i = startCol; i <= endCol; i += lowbit(i)){
            sum[endRow][i] += val;
        } */
    }

    int sum(int endRow, int endCol, int n){
        int res = 0;
        vector<int> tmpsum(n + 1);
        ++ endCol;
        ++ endRow;
        for(int i = 1; i <= endCol; ++ i){
            int val = 0;
            for(int j = endRow; j > 0; j -= lowbit(j)){
                val += sumMtx[j][i];
            }
            for(int j = i; j <= n; j += lowbit(j))
                tmpsum[j] += val;
        }
        for(int i = endCol; i > 0; i -= lowbit(i)){
            res += tmpsum[i];
        }
        return res;
    }

    int lowbit(int x){
        return x&-x;
    }
private:
vector<vector<int>> mtx;
vector<vector<int>> sumMtx;
=======
class NumMatrix {
public:
    NumMatrix(vector<vector<int>> &matrix) {
        if(matrix.size() == 0)
            return ;
        mtx.resize(matrix.size() + 1);
        sumMtx.resize(matrix.size() + 1);
        for(int i = 0; i <= matrix.size(); ++ i){
            vector<int> tmp(matrix[0].size() + 1), sumTmp(matrix[0].size() + 1);
            mtx[i] = tmp;
            sumMtx[i] = sumTmp;
        }
        for(int i = 0; i < matrix.size(); ++ i){
            for(int j = 0; j < matrix[0].size(); ++ j){
                mtx[i][j] = matrix[i][j];
                add(i + 1, j + 1, matrix.size(), matrix[0].size(), mtx[i][j]);
            }
        }
    }

    void update(int row, int col, int val) {
        if(mtx.size() == 0)
            return ;
        if(mtx[row][col] == val)
            return ;
        int delta = val - mtx[row][col];
        add(row + 1, col + 1, mtx.size() - 1, mtx[0].size(), delta);
        mtx[row][col] = val;
    }

    int sumRegion(int row1, int col1, int row2, int col2) {
        if(mtx.size() == 0)
            return 0;
        if(!row1 && !col1)
            return sum(row2, col2, mtx[0].size());
        else if(!row1)
            return sum(row2, col2, mtx[0].size()) - sum(row2, col1 - 1, mtx[0].size());
        else if(!col1)
            return sum(row2, col2, mtx[0].size()) - sum(row1 - 1, col2, mtx[0].size());
        else
            return sum(row2, col2, mtx[0].size()) - sum(row2, col1 - 1, mtx[0].size()) - sum(row1 - 1, col2, mtx[0].size()) + sum(row1 - 1, col1 - 1, mtx[0].size());
    }
private:
    void add(int startRow, int startCol, int endRow, int endCol, int val){
        for(int i = startRow; i <= endRow; i += lowbit(i)){
            sumMtx[i][startCol] += val;
        }
    /*    for(int i = startCol; i <= endCol; i += lowbit(i)){
            sum[endRow][i] += val;
        } */
    }

    int sum(int endRow, int endCol, int n){
        int res = 0;
        vector<int> tmpsum(n + 1);
        ++ endCol;
        ++ endRow;
        for(int i = 1; i <= endCol; ++ i){
            int val = 0;
            for(int j = endRow; j > 0; j -= lowbit(j)){
                val += sumMtx[j][i];
            }
            for(int j = i; j <= n; j += lowbit(j))
                tmpsum[j] += val;
        }
        for(int i = endCol; i > 0; i -= lowbit(i)){
            res += tmpsum[i];
        }
        return res;
    }

    int lowbit(int x){
        return x&-x;
    }
private:
vector<vector<int>> mtx;
vector<vector<int>> sumMtx;
>>>>>>> 6200c8704614e918c8bfa5357c648dd1b4f7eb74
};

File: algorithms/308.Range_Sum_Query_2D_-_Mutable/Range_Sum_Query_2D_-_Mutable.java
<<<<<<< HEAD
We use colSums[i][j] = the sum of ( matrix[0][j], matrix[1][j], matrix[2][j],......,matrix[i - 1][j] ).

private int[][] colSums;
private int[][] matrix;

public NumMatrix(int[][] matrix) {
    if(   matrix           == null
       || matrix.length    == 0
       || matrix[0].length == 0   ){
        return;   
     }

     this.matrix = matrix;

     int m   = matrix.length;
     int n   = matrix[0].length;
     colSums = new int[m + 1][n];
     for(int i = 1; i <= m; i++){
         for(int j = 0; j < n; j++){
             colSums[i][j] = colSums[i - 1][j] + matrix[i - 1][j];
         }
     }
}
//time complexity for the worst case scenario: O(m)
public void update(int row, int col, int val) {
    for(int i = row + 1; i < colSums.length; i++){
        colSums[i][col] = colSums[i][col] - matrix[row][col] + val;
    }

    matrix[row][col] = val;
}
//time complexity for the worst case scenario: O(n)
public int sumRegion(int row1, int col1, int row2, int col2) {
    int ret = 0;

    for(int j = col1; j <= col2; j++){
        ret += colSums[row2 + 1][j] - colSums[row1][j];
    }

    return ret;
}







public class NumMatrix {

    int[][] tree;
    int[][] nums;
    int m;
    int n;

    public NumMatrix(int[][] matrix) {
        if (matrix.length == 0 || matrix[0].length == 0) return;
        m = matrix.length;
        n = matrix[0].length;
        tree = new int[m+1][n+1];
        nums = new int[m][n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                update(i, j, matrix[i][j]);
            }
        }
    }

    public void update(int row, int col, int val) {
        if (m == 0 || n == 0) return;
        int delta = val - nums[row][col];
        nums[row][col] = val;
        for (int i = row + 1; i <= m; i += i & (-i)) {
            for (int j = col + 1; j <= n; j += j & (-j)) {
                tree[i][j] += delta;
            }
        }
    }

    public int sumRegion(int row1, int col1, int row2, int col2) {
        if (m == 0 || n == 0) return 0;
        return sum(row2+1, col2+1) + sum(row1, col1) - sum(row1, col2+1) - sum(row2+1, col1);
    }

    public int sum(int row, int col) {
        int sum = 0;
        for (int i = row; i > 0; i -= i & (-i)) {
            for (int j = col; j > 0; j -= j & (-j)) {
                sum += tree[i][j];
            }
        }
        return sum;
    }
}
=======
We use colSums[i][j] = the sum of ( matrix[0][j], matrix[1][j], matrix[2][j],......,matrix[i - 1][j] ).

private int[][] colSums;
private int[][] matrix;

public NumMatrix(int[][] matrix) {
    if(   matrix           == null
       || matrix.length    == 0
       || matrix[0].length == 0   ){
        return;   
     }

     this.matrix = matrix;

     int m   = matrix.length;
     int n   = matrix[0].length;
     colSums = new int[m + 1][n];
     for(int i = 1; i <= m; i++){
         for(int j = 0; j < n; j++){
             colSums[i][j] = colSums[i - 1][j] + matrix[i - 1][j];
         }
     }
}
//time complexity for the worst case scenario: O(m)
public void update(int row, int col, int val) {
    for(int i = row + 1; i < colSums.length; i++){
        colSums[i][col] = colSums[i][col] - matrix[row][col] + val;
    }

    matrix[row][col] = val;
}
//time complexity for the worst case scenario: O(n)
public int sumRegion(int row1, int col1, int row2, int col2) {
    int ret = 0;

    for(int j = col1; j <= col2; j++){
        ret += colSums[row2 + 1][j] - colSums[row1][j];
    }

    return ret;
}







public class NumMatrix {

    int[][] tree;
    int[][] nums;
    int m;
    int n;

    public NumMatrix(int[][] matrix) {
        if (matrix.length == 0 || matrix[0].length == 0) return;
        m = matrix.length;
        n = matrix[0].length;
        tree = new int[m+1][n+1];
        nums = new int[m][n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                update(i, j, matrix[i][j]);
            }
        }
    }

    public void update(int row, int col, int val) {
        if (m == 0 || n == 0) return;
        int delta = val - nums[row][col];
        nums[row][col] = val;
        for (int i = row + 1; i <= m; i += i & (-i)) {
            for (int j = col + 1; j <= n; j += j & (-j)) {
                tree[i][j] += delta;
            }
        }
    }

    public int sumRegion(int row1, int col1, int row2, int col2) {
        if (m == 0 || n == 0) return 0;
        return sum(row2+1, col2+1) + sum(row1, col1) - sum(row1, col2+1) - sum(row2+1, col1);
    }

    public int sum(int row, int col) {
        int sum = 0;
        for (int i = row; i > 0; i -= i & (-i)) {
            for (int j = col; j > 0; j -= j & (-j)) {
                sum += tree[i][j];
            }
        }
        return sum;
    }
}
>>>>>>> 6200c8704614e918c8bfa5357c648dd1b4f7eb74
// time should be O(log(m) * log(n))